### PR TITLE
ENG-1001: Set up deployment test suite

### PR DESCRIPTION
**Pull request by movinfinex**
It's somewhat "best effort". For example, we're currently relying on the configs being correct for some things (like council and revenue pool addresses). We can improve stuff like that as we improve the deployment processes.  (In the meantime, I guess we can manually update the configs to make tests green.)

We still need governance contract tests, and a GitHub action, and other stuff, but this is enough to start testing the concept.

### CODE REVIEW

**Comment by movinfinex** [comment]
Here's what `mainnets` looks like:

# Mainnets Report

| 🕙 Start time | ⌛ Duration |
| --- | ---: |
| 17/07/2024, 12:36:38 pm | 19.288 s |

| | ✅ Passed | ❌ Failed | ⏩ Skipped | 🚧 Todo | ⚪ Total |
| --- | ---: | ---: | ---: | ---: | ---: |
|Test Suites|2|0|0|0|8|
|Tests|115|5|10|0|130|

## ✅ <a id="file0" href="#file0">tests/craterun.test.ts</a> [[link](https://github.com/infinex-xyz/infinex-deployments/test/tests/craterun.test.ts)]

10 passed, 0 failed, 0 skipped, 0 todo, done in 1747 s

```
✅ Craterun (mainnets)
   ✅ Opener is a deployed contract
   ✅ Opener owner matches config
   ✅ Opener has no pending owner
   ✅ Opener crates seem valid
   ✅ Points proxy is a deployed contract
   ✅ Points proxy points name configured
   ✅ CTP staking rewards is a deployed contract
   ✅ Opener recognises keeper from Platform app
   ✅ CTP staking rewards recognises keeper
   ✅ Pyth Entropy configuration looks valid
```

## ❌ <a id="file1" href="#file1">tests/multichain.test.ts</a> [[link](https://github.com/infinex-xyz/infinex-deployments/test/tests/multichain.test.ts)]

105 passed, 5 failed, 10 skipped, 0 todo, done in 14847 s

```
❌ Multichain (base mainnets)
   ✅ Account factory is a deployed contract
   ✅ Account factory can predict address
   ✅ Account factory can create account
   ❌ Account factory has been promoted to latest
   ✅ Account factory uses latest beacon
   ✅ Forwarder is a deployed contract
   ✅ Forwarder is valid
   ⏩ Accounts router v1 is a deployed contract
   ⏩ Accounts router v2 is a deployed contract
   ✅ Protocol config beacon is a deployed contract
   ✅ Protocol config beacon has revenue pool from config
   ✅ Protocol config beacon points to valid USDC contract
   ✅ Protocol config beacon points to valid app registry contract
   ✅ Protocol config beacon recognises recovery keeper address from Platform app
   ✅ Protocol config beacon has been promoted to latest
   ✅ App registry is a deployed contract
   ✅ App registry owner matches config
   ✅ App registry has no pending owner
   ✅ App registry app beacon interface supported by curve stableswap app beacon
   ✅ Curve stableswap app beacon is a deployed contract
   ✅ Curve stableswap app beacon owner matches config
   ✅ Curve stableswap app beacon has no pending owner
   ✅ Curve stableswap app beacon points to valid USDC contract
   ✅ Curve stableswap app beacon is recognised by app registry
❌ Multichain (arbitrum mainnets)
   ✅ Account factory is a deployed contract
   ✅ Account factory can predict address
   ✅ Account factory can create account
   ❌ Account factory has been promoted to latest
   ✅ Account factory uses latest beacon
   ✅ Forwarder is a deployed contract
   ✅ Forwarder is valid
   ⏩ Accounts router v1 is a deployed contract
   ⏩ Accounts router v2 is a deployed contract
   ✅ Protocol config beacon is a deployed contract
   ✅ Protocol config beacon has revenue pool from config
   ✅ Protocol config beacon points to valid USDC contract
   ✅ Protocol config beacon points to valid app registry contract
   ✅ Protocol config beacon recognises recovery keeper address from Platform app
   ✅ Protocol config beacon has been promoted to latest
   ✅ App registry is a deployed contract
   ✅ App registry owner matches config
   ✅ App registry has no pending owner
   ✅ App registry app beacon interface supported by curve stableswap app beacon
   ✅ Curve stableswap app beacon is a deployed contract
   ✅ Curve stableswap app beacon owner matches config
   ✅ Curve stableswap app beacon has no pending owner
   ✅ Curve stableswap app beacon points to valid USDC contract
   ✅ Curve stableswap app beacon is recognised by app registry
❌ Multichain (optimism mainnets)
   ✅ Account factory is a deployed contract
   ✅ Account factory can predict address
   ✅ Account factory can create account
   ❌ Account factory has been promoted to latest
   ✅ Account factory uses latest beacon
   ✅ Forwarder is a deployed contract
   ✅ Forwarder is valid
   ⏩ Accounts router v1 is a deployed contract
   ⏩ Accounts router v2 is a deployed contract
   ✅ Protocol config beacon is a deployed contract
   ✅ Protocol config beacon has revenue pool from config
   ✅ Protocol config beacon points to valid USDC contract
   ✅ Protocol config beacon points to valid app registry contract
   ✅ Protocol config beacon recognises recovery keeper address from Platform app
   ✅ Protocol config beacon has been promoted to latest
   ✅ App registry is a deployed contract
   ✅ App registry owner matches config
   ✅ App registry has no pending owner
   ✅ App registry app beacon interface supported by curve stableswap app beacon
   ✅ Curve stableswap app beacon is a deployed contract
   ✅ Curve stableswap app beacon owner matches config
   ✅ Curve stableswap app beacon has no pending owner
   ✅ Curve stableswap app beacon points to valid USDC contract
   ✅ Curve stableswap app beacon is recognised by app registry
❌ Multichain (polygon mainnets)
   ✅ Account factory is a deployed contract
   ✅ Account factory can predict address
   ✅ Account factory can create account
   ❌ Account factory has been promoted to latest
   ✅ Account factory uses latest beacon
   ✅ Forwarder is a deployed contract
   ✅ Forwarder is valid
   ⏩ Accounts router v1 is a deployed contract
   ⏩ Accounts router v2 is a deployed contract
   ✅ Protocol config beacon is a deployed contract
   ✅ Protocol config beacon has revenue pool from config
   ✅ Protocol config beacon points to valid USDC contract
   ✅ Protocol config beacon points to valid app registry contract
   ✅ Protocol config beacon recognises recovery keeper address from Platform app
   ✅ Protocol config beacon has been promoted to latest
   ✅ App registry is a deployed contract
   ✅ App registry owner matches config
   ✅ App registry has no pending owner
   ✅ App registry app beacon interface supported by curve stableswap app beacon
   ✅ Curve stableswap app beacon is a deployed contract
   ✅ Curve stableswap app beacon owner matches config
   ✅ Curve stableswap app beacon has no pending owner
   ✅ Curve stableswap app beacon points to valid USDC contract
   ✅ Curve stableswap app beacon is recognised by app registry
❌ Multichain (ethereum mainnets)
   ✅ Account factory is a deployed contract
   ✅ Account factory can predict address
   ✅ Account factory can create account
   ❌ Account factory has been promoted to latest
   ✅ Account factory uses latest beacon
   ✅ Forwarder is a deployed contract
   ✅ Forwarder is valid
   ⏩ Accounts router v1 is a deployed contract
   ⏩ Accounts router v2 is a deployed contract
   ✅ Protocol config beacon is a deployed contract
   ✅ Protocol config beacon has revenue pool from config
   ✅ Protocol config beacon points to valid USDC contract
   ✅ Protocol config beacon points to valid app registry contract
   ✅ Protocol config beacon recognises recovery keeper address from Platform app
   ✅ Protocol config beacon has been promoted to latest
   ✅ App registry is a deployed contract
   ✅ App registry owner matches config
   ✅ App registry has no pending owner
   ✅ App registry app beacon interface supported by curve stableswap app beacon
   ✅ Curve stableswap app beacon is a deployed contract
   ✅ Curve stableswap app beacon owner matches config
   ✅ Curve stableswap app beacon has no pending owner
   ✅ Curve stableswap app beacon points to valid USDC contract
   ✅ Curve stableswap app beacon is recognised by app registry
```

**Comment by movinfinex** [comment]
This is what the web UI looks like:

Screenshot 2024-07-17 at 1 14 23 pm

File: test/lib/utils.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
+import { evmChains, type EvmChains } from "@infinex/infinex-sdk/sdk/types";
+import { ethers } from "ethers";
+import * as toml from "smol-toml";
+import fs from "node:fs/promises";
+import dotenv from "dotenv";
+
+export { evmChains, type EvmChains } from "@infinex/infinex-sdk/sdk/types";
+export { ethers } from "ethers";
```
Comment (yingli-infinex) [inline test/lib/utils.ts:8]:
```
export { evmChains, type EvmChains } from "@infinex/infinex-sdk/sdk/types";
export { ethers } from "ethers";
```

Duplicate

**Review by yingli-infinex — COMMENTED** [review]
@axefinaxe This looks great to me, anything you think we should add? Im not 100% sure what the deployments are doing anymore, in addition what configuration is considered "valid" or "safe". 

@movinfinex Just one comment, about duplicate code. Other than that I think its a great idea to have a config smoke test during staging and prod deployments

File: test/lib/utils.ts (hunk 1 above)
Comment (movinfinex) [reply test/lib/utils.ts:8]: Do mean with lines 1 and 2?  These are exports and they're imports.

**Review by rayInfinex — APPROVED** [review]
Seems like a good intermediary test to have. I don't have any issues with adding this.

**Pull request merged**